test_python/pytorch_example.py

```diff
@@ -0,0 +1,527 @@
+import time
+import torch
+
+dump_backward_overhead = False
+
+################################################################################
+# 0. Initializations
+################################################################################
+from tensor_comprehensions.tclib import MappingOptions
+
+# Define a timing function to print some results
+def time_tc(iters, prepend, runFun, tc_name, inputs):
+    global dump_backward_overhead
+
+    timesCPU = []
+    timesCPUAndGPU = []
+    for i in range(iters):
+        torch.cuda.synchronize()
+        start = time.clock()
+        if dump_backward_overhead:
+            dump_backward_overhead = time.clock()
+        outputs = runFun(tc_name, inputs, ())
+        timesCPU.append(time.clock() - start)
+        torch.cuda.synchronize()
+        timesCPUAndGPU.append(time.clock() - start)
+    print("#################################################################")
+    timesCPU = sorted(timesCPU)
+    print("{} Total CPU time to launch kernel: min {}us, p50 {}us, p90 {}us, max {}us".format(
+        prepend,
+        int(timesCPU[0] * 1e6),
+        int(timesCPU[int(len(timesCPU) // 2)] * 1e6),
+        int(timesCPU[int((len(timesCPU) * 9) // 10)] * 1e6),
+        int(timesCPU[len(timesCPU) - 1] * 1e6),
+    ))
+    timesCPUAndGPU = sorted(timesCPUAndGPU)
+    print("{} Total CPU launch + GPU kernel time: min {}us, p50 {}us, p90 {}us, max {}us".format(
+        prepend,
+        int(timesCPUAndGPU[0] * 1e6),
+        int(timesCPUAndGPU[int(len(timesCPUAndGPU) // 2)] * 1e6),
+        int(timesCPUAndGPU[int((len(timesCPUAndGPU) * 9) // 10)] * 1e6),
+        int(timesCPUAndGPU[len(timesCPUAndGPU) - 1] * 1e6),
+    ))
+
+# Define a TC string for matmul and some input torch cuda tensors
+mm = """
+def matmul(float(M,N) A, float(N,K) B) -> (C) {
+    C(m, k) +=! A(m, r_n) * B(r_n, k)
+}
+def matmul_grad(float(M,N) A, float(N,K) B, float(M,K) d_O) -> (d_A, d_B) {
+    d_A(m, n) +=! d_O(  m, r_k) * B(  n, r_k)
+    d_B(n, k) +=! d_O(r_m,   k) * A(r_m,   n)
+}
+"""
+mat1, mat2 = torch.randn(300, 400).cuda(), torch.randn(400, 500).cuda()
+
+################################################################################
+# 1. Use the C++ API to build a low-overhead compilation cache and time it
+################################################################################
+from tensor_comprehensions.tclib import CompilationCache
+
+compilation_cache = CompilationCache(mm)
```

Comment:
can we make a minimal example without compilation cache?

Reply:
The idea in this file is to use the python bindings to build python abstractions.
But this should not be the entry point for a user.
I'll rename the file and make it clear that this is internal.
The next PR will reimplement the user-facing API with these primitives.

Reply:
yes, the filename got me confused, it implies the the code is an example of using tc in pytorch